Sources/LeafKit/LeafData/LeafDataStorage.swift

```diff
@@ -3,46 +3,220 @@
+              returns: LeafData.NaturalType,
+              invariant: Bool)
+    
+    // MARK: - LeafSymbol Conformance
+    
+    // MARK: Properties
+    internal var resolved: Bool { true }
+    internal var invariant: Bool {
+        switch self {
+            case .lazy(_, _, let invariant): return invariant
+            case .optional(let o, _): return o?.invariant ?? true
+            case .array(let a):
+                let stored = a.map { $0.storage }.filter { $0.isLazy }
+                return stored.allSatisfy { $0.invariant }
+            case .dictionary(let d):
+                let stored = d.values.map { $0.storage }.filter { $0.isLazy }
+                return stored.allSatisfy { $0.invariant }
+            default: return true
+        }
+    }
+    internal var symbols: Set<String> { .init() }
+    internal var isAtomic: Bool { true }
+    internal var isExpression: Bool { false }
+    internal var isAny: Bool { false }
+    internal var isConcrete: Bool { true }
+    /// Note: Will *always* return a value - can be force-unwrapped safely
+    internal var concreteType: LeafData.NaturalType? {
+        switch self {
+            // Concrete Types
+            case .array(_)           : return .array
+            case .bool(_)            : return .bool
+            case .data(_)            : return .data
+            case .dictionary(_)      : return .dictionary
+            case .double(_)          : return .double
+            case .int(_)             : return .int
+            case .string(_)          : return .string
+            // Internal Wrapped Types
+            case .lazy(_, let t, _),
+                 .optional(_, let t) : return t
+        }
+    }
+    
+    // MARK: Functions
+    
+    /// Will resolve anything but variant Lazy data (99% of everything), and unwrap optionals
+    internal func softResolve() -> LeafDataStorage {
+        guard invariant else { return self }
+        switch self {
+            case .lazy(let f, _, _): return f().storage
+            case .optional(let o, _):
+                if let unwrapped = o { return unwrapped }
+                return self
+            case .array(let a):
+                let resolved: [LeafData] = a.map {
+                    LeafData($0.storage.softResolve())
+                }
+                return .array(resolved)
+            case .dictionary(let d):
+                let resolved: [String: LeafData] = d.mapValues {
+                    LeafData($0.storage.softResolve())
+                }
+                return .dictionary(resolved)
+            default: return self
+        }
+    }
+
+    /// Will resolve anything but variant Lazy data (99% of everything)
+    internal func hardResolve() throws -> LeafData {
+        let resolveStored: LeafDataStorage = self.softResolve()
+        return LeafData(resolveStored)
+    }
+    
+    /// Will serialize anything to a String except Lazy -> Lazy
+    internal func serialize() throws -> String? {
```

Comment:
Future LeafSymbol adherence but also the final step to serializing a template. This can *potentially* be called prior to serialization to compress templates when data can be determined to no longer be variant.

EG: `#localizedCurrency(1_000, "en_us")` would be a forced coercion of a variable from .int to .string, and would be pointless to store as `tag(localizedCurrency, [1000], "en_us")` when it could just be `.raw("$1,000.00")`